homeassistant/components/music_assistant/config_flow.py

```diff
@@ -0,0 +1,168 @@
+
+
+def get_manual_schema(user_input: dict[str, Any]) -> vol.Schema:
+    """Return a schema for the manual step."""
+    default_url = user_input.get(CONF_URL, DEFAULT_URL)
+    return vol.Schema(
+        {
+            vol.Required(CONF_URL, default=default_url): str,
+        }
+    )
+
+
+async def get_server_info(hass: HomeAssistant, url: str) -> ServerInfoMessage:
+    """Validate the user input allows us to connect."""
+    async with MusicAssistantClient(
+        url, aiohttp_client.async_get_clientsession(hass)
+    ) as client:
+        if TYPE_CHECKING:
+            assert client.server_info is not None
+        return client.server_info
+
+
+# ruff: noqa: ARG002
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for MusicAssistant."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Set up flow instance."""
+        self.server_info: ServerInfoMessage | None = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        return await self.async_step_manual()
+
+    async def async_step_manual(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a manual configuration."""
+        if user_input is None:
+            return self.async_show_form(  # type: ignore[no-any-return]
+                step_id="manual", data_schema=get_manual_schema({})
+            )
+
+        errors = {}
+        try:
+            self.server_info = await get_server_info(self.hass, user_input[CONF_URL])
+            await self.async_set_unique_id(self.server_info.server_id)
+        except CannotConnect:
+            errors["base"] = "cannot_connect"
+        except InvalidServerVersion:
+            errors["base"] = "invalid_server_version"
+        except MusicAssistantClientException:  # pylint: disable=broad-except
+            LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+        else:
+            return await self._async_create_entry_or_abort()
+
+        return self.async_show_form(  # type: ignore[no-any-return]
+            step_id="manual", data_schema=get_manual_schema(user_input), errors=errors
+        )
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> ConfigFlowResult:
+        """Handle a discovered Mass server.
+
+        This flow is triggered by the Zeroconf component. It will check if the
+        host is already configured and delegate to the import step if not.
+        """
+        # abort if discovery info is not what we expect
+        if "server_id" not in discovery_info.properties:
+            return self.async_abort(reason="missing_server_id")  # type: ignore[no-any-return]
+        # abort if we already have exactly this server_id
+        # reload the integration if the host got updated
```

Comment:
should we only accept the updated host if we're in a failed state ?
As in: if the user configured some custom ip/dns name and that is connected, why would we override it if the connection is working.